jetbrains-core/src/software/aws/toolkits/jetbrains/core/experiments/ToolkitExperiment.kt

```diff
@@ -89,14 +137,40 @@ internal class ToolkitExperimentManager : PersistentStateComponent<ExperimentSta
         }
     }
 
+    internal fun shouldPrompt(experiment: ToolkitExperiment, now: Instant = Instant.now()): Boolean {
+        if (experiment.isEnabled()) {
+            return false
+        }
+        val should = state.nextSuggestion[experiment.id]?.let { it > 0 && now.isAfter(Instant.ofEpochMilli(it)) } ?: true
+        if (should) {
+            state.nextSuggestion[experiment.id] = now.plus(experiment.suggestionSnooze).toEpochMilli()
+        }
+        return should
+    }
+
+    internal fun neverPrompt(experiment: ToolkitExperiment) {
+        state.nextSuggestion[experiment.id] = -1
+    }
+
     companion object {
         internal val EP_NAME = ExtensionPointName.create<ToolkitExperiment>("aws.toolkit.experiment")
+        internal val EXPERIMENT_CHANGED = Topic.create("experiment service enable state changed", ToolkitExperimentStateChangedListener::class.java)
         internal fun getInstance(): ToolkitExperimentManager = service()
         internal fun visibleExperiments(): List<ToolkitExperiment> = EP_NAME.extensionList.filterNot { it.hidden }
     }
 }
 
 internal class ExperimentState : BaseState() {
+    // This represents whether an experiment is enabled or not, don't want to rename it as that will
+    // cause problems with any experiments already out there in the wild who've been persisted
+    // as 'value'
     @get:Property
     val value by map<String, Boolean>()
+
+    @get:Property
+    val nextSuggestion by map<String, Long>()
```

Comment:
can we keep this as Instant instead of converting to epoch millis?

Reply:
the `BaseState` serialization doesn't really support it out of the box - and the nice thing about this is we can represent "never" as `-1`

I can convert if you think necessary tho, the `-1` thing is just a convenience.